### PR TITLE
Feature/sagemaker

### PR DESCRIPTION
**Pull request by vanpelt**
I moved the tensorflow hook into `wandb.tensorflow`.  This also adds support for sagemaker and replaces `hook_torch` with `watch`.

### CODE REVIEW

**Comment by vanpelt** [comment]
@shawn this fixes the memory reporting as well!   Feel free to merge and test Luke’s stuff as well.

File: wandb/__init__.py [hunk 1]
```diff
@@ -456,6 +464,26 @@ def _get_python_type():
         return "python"
 
 
+def parse_sm_config():
+    sagemaker_config = "/opt/ml/input/config/hyperparameters.json"
+    if os.path.exists(sagemaker_config):
+        conf = {}
+        # Hyper-parameter searchs quote configs...
+        for k, v in six.iteritems(json.load(open(sagemaker_config))):
```
Comment (shawnlewis) [inline wandb/__init__.py]: Can the sagemaker config have nested objects and arrays? (In which case we'd want to do this recursively?)

File: wandb/__init__.py [hunk 2]
```diff
@@ -483,6 +511,17 @@ def init(job_type=None, dir=None, config=None, project=None, entity=None, group=
                            "WANDB_PROJECT", "WANDB_API_KEY"])
         run = None
 
+    sagemaker_config = parse_sm_config()
+    if sagemaker_config:
+        # Set run_id and potentially grouping if we're in SageMaker
+        os.environ['WANDB_RUN_ID'] = '-'.join([
```
Comment (shawnlewis) [inline wandb/__init__.py]: If TRAINING_JOB_NAME and CURRENT_HOST are not available we'll always use the same run ID which doesn't seem great.

File: wandb/__init__.py [hunk 3]
```diff
@@ -483,6 +511,17 @@ def init(job_type=None, dir=None, config=None, project=None, entity=None, group=
...
+            os.getenv('TRAINING_JOB_NAME', 'sagemaker'),
+            os.getenv('CURRENT_HOST', 'algo1')])
+        conf = json.load(
+            open("/opt/ml/input/config/resourceconfig.json"))
+        if group == None and len(conf["hosts"]) > 1:
```
Comment (shawnlewis) [inline wandb/__init__.py]: Pretty sweet that we'd automatically work with distributed jobs.

File: wandb/stats.py [hunk 4]
```diff
@@ -81,11 +84,15 @@ def __init__(self, run, api):
...
-            "recv": net.bytes_recv
-        }
+        if psutil:
+            net = psutil.net_io_counters()
+            self.network_init = {
+                "sent": net.bytes_sent,
+                "recv": net.bytes_recv
+            }
+        else:
+            termlog(
```
Comment (shawnlewis) [inline wandb/stats.py]: But we still keep it as a pip dependency for the main wandb package right?

File: wandb/stats.py [hunk 5]
```diff
@@ -151,28 +158,30 @@ def stats(self):
             handle = nvmlDeviceGetHandleByIndex(i)
             try:
                 util = nvmlDeviceGetUtilizationRates(handle)
+                memory = nvmlDeviceGetMemoryInfo(handle)
                 temp = nvmlDeviceGetTemperature(handle, NVML_TEMPERATURE_GPU)
                 stats["gpu.{0}.{1}".format(i, "gpu")] = util.gpu
-                stats["gpu.{0}.{1}".format(i, "memory")] = util.memory
+                stats["gpu.{0}.{1}".format(i, "memory")] = memory.used
```
Comment (shawnlewis) [inline wandb/stats.py]: I think we might need to use a different key unfortunately. The UI needs to be able to display the correct label which depends on which CLI version sent it.

An alternative would be to store the CLI version on each run in the DB, then the UI could switch on that.

File: wandb/__init__.py (hunk 1 above)
Comment (vanpelt) [reply wandb/__init__.py]: I don't think they can, but they are doing a weird json encoding thing I wish I understood.

**Pull request merged**